Framework/API/src/Algorithm.cpp

```diff
@@ -1651,6 +1656,31 @@ bool Algorithm::getAlgStartupLogging() const {
   return m_isAlgStartupLoggingEnabled;
 }
 
+void Algorithm::declareProperty(std::unique_ptr<Kernel::Property> p,
+                                const std::string &doc) {
+  if (isCompoundProperty(p->name()))
+    throw std::runtime_error(
+        p->name() +
+        " has already been used by Algorithm::declareIndexProperty.");
+
+  PropertyManagerOwner::declareProperty(std::move(p), doc);
+}
+
+Kernel::IPropertyManager::TypedValue
+Algorithm::getProperty(const std::string &name) const {
+  if (isCompoundProperty(name))
+    throw std::runtime_error("Algorithm::getIndexProperty must be used with "
+                             "properties declared using "
+                             "Algorithm::declareIndexProperty.");
+
+  return PropertyManagerOwner::getProperty(name);
+}
+
+bool Algorithm::isCompoundProperty(const std::string &name) const {
+  return std::find(m_reservedList.cbegin(), m_reservedList.cend(), name) !=
```

Comment:
I don't think there's anything you can/should programatically do. My concern is that for ease of development, it's going to be better practice for new algorithms to make the `declareIndexProperty` calls first, because that's where the reserved list gets populated. For existing algorithms, this could cause some issues because the order of declaration defines the order of properties in the GUI dialogs (AlgorithmDialog) and dictates the layout for python functions generated in the `simpleapi`. I think this is something we might want to consider as we get closer to a rollout.